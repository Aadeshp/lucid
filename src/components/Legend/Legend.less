.lucid-Legend {
	display: inline-flex;
	color: @color-darkGray;
	font-size: @fontSize;
	list-style-type: none;
	margin: 0;
	padding: 0;
	background-color: @color-white;

	&-is-horizontal {
		flex-direction: row;
		height: @size-standard-height;
		padding: 0 @size-S;
	}
	&-is-vertical {
		flex-direction: column;
	}
	&-has-borders {
		border: @border-lightBorder;
		border-radius: @size-borderRadius;
	}
	&-is-vertical&-has-borders {
		padding: @size-S;
	}

	&-Item {
		display: flex;
		align-items: center;
	}
	&-is-vertical &-Item {
		padding-top: @size-S;
	}
	&-is-horizontal &-Item {
		padding-left: @size-L;
	}
	&-Item:first-of-type {
		padding-left: 0;
		padding-top: 0;
	}

	&-Item-indicator {
		flex-shrink: 0;
		margin-right: @size-XS;
	}
}
