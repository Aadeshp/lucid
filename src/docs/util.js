import _ from 'lodash';
import { markdown } from 'markdown';

export function toMarkdown(md) {
	if (md) {
		return {
			__html: markdown.toHTML(md),
		};
	}

	return {
		__html: '',
	};
}

export function handleHighlightCode() {
	if (window.hljs) {
		//eslint-disable-line
		_.each(document.querySelectorAll('pre code'), block => {
			hljs.highlightBlock(block); //eslint-disable-line
		});
	}
}

export function sanitizeExamplePath(path) {
	return _.flow(
		x => _.split(x, '/'),
		x => _.compact(x),
		x => _.drop(x),
		x => _.join(x, '-'),
		x => _.split(x, '.'),
		x => _.join(x, '-')
	)(path);
}

export function stripIndent(text) {
	if (!text) {
		return '';
	}

	const sigLines = text.split(/\n/).filter(line => !/^\s*$/.test(line));
	const leadingWhitespaces = sigLines.map(line => /^(\s*)/g.exec(line)[1]);
	const minimumLeadingWhitespace = leadingWhitespaces.reduce(
		(min, line) => (line.length < min.length ? line : min),
		leadingWhitespaces[0]
	);
	return text
		.split(/\n/)
		.map(line => line.replace(minimumLeadingWhitespace, ''))
		.join('\n');
}
