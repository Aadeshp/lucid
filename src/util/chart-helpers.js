import _ from 'lodash';
import d3TimeFormat from 'd3-time-format';

/**
 * groupByFields
 *
 * This will return the data in a similar format to d3Shape.stack
 * but without the stacking of the data.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function groupByFields(collection, fields) {
	const fieldsArray = [].concat(fields);

	return _.map(fieldsArray, (field) => {
		return _.map(collection, (item) => item[field]);
	});
}

/**
 * toPairs
 *
 * Takes a collection of data and pairs up some x and y data.
 *
 * @param {object[]} collection
 * @param {string} xField
 * @param {string} yField
 * @return {array} - E.g. [[x, y], [x, y]]
 */
export function toPairs(collection, xField, yField) {
	return _.map(collection, (item) => [item[xField], item[yField]]);
}

/**
 * byFields
 *
 * Takes a collection of data and returns an array of all the fields off that
 * collection.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {array}
 */
export function byFields(collection, fields) {
	const fieldArray = [].concat(fields);

	return _.reduce(fieldArray, (acc, field) => {
		return acc.concat(_.map(collection, field));
	}, []);
}

/**
 * minByFields
 *
 * Returns the minimum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function minByFields(collection, fields) {
	return _.min(byFields(collection, fields));
}

/**
 * maxByFields
 *
 * Returns the maximum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFields(collection, fields) {
	return _.max(byFields(collection, fields));
}

/**
 * maxByFieldsStacked
 *
 * Returns the max sum of a set of fields from a collection
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFieldsStacked(collection, fields) {
	const fieldArray = [].concat(fields);

	const sums = _.reduce(collection, (acc, item) => {
		const sum = _.chain(item)
			.pick(fieldArray)
			.toArray()
			.sum()
			.value();
		return acc.concat(sum);
	}, []);

	return _.max(sums);
}

/**
 * discreteTicks
 *
 * Returns `count` evenly spaced, representative values from the `array`.
 *
 * @param {array} array
 * @param {number} size - should be greater than 1
 * @return {array}
 */
export function discreteTicks(array, count) {
	if (!array || _.isNil(count) || array.length <= count) {
		return array;
	}

	const step = (array.length - 1) / Math.max(1, count - 1);

	return _.reduce(_.times(count), (acc, n) => {
		return acc.concat(array[Math.round(n  * step)]);
	}, []);
}

/**
 * transformFromCenter
 *
 * Scaling paths from their center is tricky. This function
 * helps do that be generating a translate/scale transform
 * string with the correct numbers.
 *
 * @param {number} x - the x data point where you want the path to be centered at
 * @param {number} y - the y data point where you want the path to be centered at
 * @param {number} xCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} yCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} scale - number to scale to, 2 would be 2x bigger
 * @return {string} - transform string
 */
export function transformFromCenter(x, y, xCenter, yCenter, scale) {
	return `translate(${((1 - scale) * xCenter) + (x - xCenter)}, ${((1 - scale) * yCenter) + (y - yCenter)}) scale(${scale})`;
}

const formatMillisecond = d3TimeFormat.timeFormat('.%L');
const formatSecond = d3TimeFormat.timeFormat(':%S');
const formatMinute = d3TimeFormat.timeFormat('%I:%M');
const formatHour = d3TimeFormat.timeFormat('%I %p');
const formatDay = d3TimeFormat.timeFormat('%a %d');
const formatWeek = d3TimeFormat.timeFormat('%b %d');
const formatMonth = d3TimeFormat.timeFormat('%B');
const formatYear = d3TimeFormat.timeFormat('%Y');

export function multiFormat(date) {
	return (d3TimeFormat.timeSecond(date) < date ? formatMillisecond
		: d3TimeFormat.timeMinute(date) < date ? formatSecond
		: d3TimeFormat.timeHour(date) < date ? formatMinute
		: d3TimeFormat.timeDay(date) < date ? formatHour
		: d3TimeFormat.timeMonth(date) < date ? (d3TimeFormat.timeWeek(date) < date ? formatDay : formatWeek)
		: d3TimeFormat.timeYear(date) < date ? formatMonth
		: formatYear)(date);
}

