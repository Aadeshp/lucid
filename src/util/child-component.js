import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {
			isRenderNull: true,

			/**
			 * findInChildren
			 *
			 * Returns all elements with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} elements with this component type
			 */
			findInChildren(elements) {
				let component = this;
				let ownElements = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						ownElements.push(element);
					}
				});
				return ownElements;
			},

			/**
			 * findInChildrenAsProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} `props` objects of elements with this component type
			 */
			findInChildrenAsProps(elements) {
				let component = this;
				return _.map(component.findInChildren(elements), 'props');
			},

			/**
			 * findInProps
			 *
			 * Given a set of parent props this will return only the props for this
			 * component type
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of normalized props
			 */
			findInProps(parentProps) {
				// Extract out the defined `childProps` from the parent class
				const currentProps = _.chain(parentProps)
					.get(definition.propName) // grab the prop we care about, e.g. "Child"
					.thru(x => [x]) // wrap in an array
					.flatten()
					.filter(x => x) // remove falsey values
					.value();

				return _.map(currentProps, (childProp) => {
					if (_.isPlainObject(childProp)) {
						return childProp;
					} else {
						return {
							children: childProp
						};
					}
				});
			},

			/**
			 * findInAllAsProps
			 *
			 * Given a parent props object, this will return a mashup of all the
			 * children and/or props for the current component type. This is
			 * particularly useful for providing a flexible component API. It allows
			 * consumers to either pass stuff through props or children.
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of `props`
			 */
			findInAllAsProps(parentProps) {
				let component = this;

				const propsFromProps = component.findInProps(parentProps);
				const propsFromChildren = component.findInChildrenAsProps(parentProps.children);

				return propsFromProps.concat(propsFromChildren);
			}
		}
	});
}

export function rejectNullElements(children) {
	return _.reject(React.Children.toArray(children), (node) => (React.isValidElement(node) && node.type.isRenderNull));
}
