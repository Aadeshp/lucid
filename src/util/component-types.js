import React from 'react';
import _ from 'lodash';

// creates a React component
export function createClass(definition={}) {
	const {
		_isPrivate = false,
		statics = {},
		components = {},
		reducers = {},
		selectors = {},
		propName = null,
		propTypes = {},
		render = () => null,
		...restDefinition
	} = definition;

	const newDefinition = {
		...restDefinition,
		statics: {
			...statics,
			...components,
			_isPrivate,
			reducers,
			selectors,
			propName,
		},
		propTypes: _.assign({}, propTypes, _.mapValues(definition.components, () => React.PropTypes.any)),
		render,
	};

	newDefinition.statics.definition = newDefinition;

	return React.createClass(newDefinition);
}

// return all elements matching the specified types
export function filterTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.filter(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return all elements not matching the specified types
export function rejectTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.reject(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return an array of elements (of the given type) for each of the values
export function createElements(type, values=[]) {
	return _.reduce([].concat(values), (elements, typeValue) => {
		if (React.isValidElement(typeValue) && typeValue.type === type) {
			return elements.concat(typeValue);
		} else if (_.isPlainObject(typeValue) && !React.isValidElement(typeValue)) {
			return elements.concat(React.createElement(type, typeValue));
		} else if (_.isUndefined(typeValue)) {
			return elements;
		} else {
			return elements.concat(React.createElement(type, null, typeValue));
		}
	}, []);
}

// return all elements found in props and children of the specified types
export function findTypes(props, types=[]) {
	// get elements from props (using type.propName)
	const elementsFromProps = _.reduce(_.castArray(types), (acc, type) => {
		if (!_.isUndefined(type.propName)) {
			const propMatches = _.flatten(_.values(_.pick(props, type.propName)));
			return acc.concat(createElements(type, propMatches));
		}
		return acc;
	}, []);

	// return elements from props and elements from children
	return elementsFromProps.concat(filterTypes(props.children, types));
}

// return the first element found in props and children of the specificed type(s)
export function getFirst(props, types, defaultValue) {
	return _.first(findTypes(props, types)) || defaultValue;
}

// omit props defined in propTypes of the given type
// and any extra keys given in third argument
export function omitProps(props, type, keys = []) {
	return _.omit(props, _.keys(type.propTypes).concat(keys).concat(['initialState']));
}
