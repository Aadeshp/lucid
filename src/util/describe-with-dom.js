function describeWithDOM(name, func) {
	describe(`${name} (functional tests using ${window.realOrFake})`, func);
}

describeWithDOM.only = function (name, func) {
	describe.only(`${name} (functional tests using ${window.realOrFake})`, func);
};

export default describeWithDOM;
