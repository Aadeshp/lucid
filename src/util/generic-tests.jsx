import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, getDefaultProps=_.noop) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe(`[common]`, () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f'
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			assert.deepEqual(wrapper.first().prop('style'), style);
		});

		it('should pass through `className`', () => {
			let expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			let expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let parentClasses = wrapper.first().prop('className').split(' ');
			let childrenClasses = wrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			let allClasses = parentClasses.concat(childrenClasses);

			assert(_.every(allClasses, (className) => {
				return _.includes(className, 'lucid-' + Component.displayName);
			}));
		});

		it('should only use onX convention for function proptypes', () => {
			assert(_.every(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				let isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				let isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`
				if (isProbablyFunction && !isAny && !_.startsWith(key, 'on')) {
					return false
				}

				return true;
			}));
		});

		it('should be available as an exported module from index.js', () => {
			assert(lucid[Component.displayName]);
		});
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM(`[icon]`, () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `uniqueId` and `event` prop', () => {
			const expectedUniqueId = 32;
			const props = {
				uniqueId: expectedUniqueId,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { uniqueId, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(uniqueId, expectedUniqueId, 'incorrect or missing uniqueId');
		});
	});
}
