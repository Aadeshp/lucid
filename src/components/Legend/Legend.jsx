import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass, findTypes } from '../../util/component-types';

import Point from '../Point/Point';
import Line from '../Line/Line';

const cx = lucidClassNames.bind('&-Legend');

const POINT_SIZE = 12;
const LINE_WIDTH = 22;

const {
	number,
	string,
	oneOf,
	bool,
	func,
} = React.PropTypes;

/**
 * {"categories": ["visualizations", "chart primitives"]}
 *
 * Contrary to the other chart primitives, this component is not rendered in
 * svg. In order to sanely render horizontal legends, we need to know the width
 * of the text elements ahead of rendering time. Since we're using a variable
 * width font, the only way I know of to correctly get the width is with the
 * DOM function `getComputedTextLength`. Variable widths are much more easy to
 * implement outside of svg.
 */
const Legend = createClass({
	displayName: 'Legend',

	statics: {
		HEIGHT: 28, // exposed for consumer convenience
	},

	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root element.
		 */
		className: string,
		/**
		 * Determine orientation of the legend.
		 */
		orient: oneOf(['horizontal', 'vertical']),
		/**
		 * Show the legend borders. Turn this off if you want to put the legend in
		 * a `ToolTip` for example.
		 */
		hasBorders: bool,
	},

	getDefaultProps() {
		return {
			orient: 'vertical',
			hasBorders: true,
		};
	},

	components: {
		Item: createClass({
			displayName: 'Legend.Item',
			propsName: 'Item',
			propTypes: {
				hasPoint: bool,
				hasLine: bool,
				/**
				 * Strings should match an existing color class unless they start with a
				 * '#' for specific colors. E.g.:
				 *
				 * - `COLOR_0`
				 * - `COLOR_GOOD`
				 * - `'#123abc'`
				 */
				color: string,
				pointKind: number,
				onClick: func,
			},
		}),
	},

	handleItemClick(index, props, event) {
		if (!props.onClick) {
			return null;
		}

		props.onClick(index, { props, event });
	},

	render() {
		const {
			orient,
			className,
			hasBorders,
			...passThroughs,
		} = this.props;

		const isHorizontal = orient === 'horizontal';
		const isVertical = orient === 'vertical';
		const itemProps = _.map(findTypes(this.props, Legend.Item), 'props');
		const hasSomeLines = isVertical && _.some(itemProps, ({ hasLine }) => hasLine);

		return (
			<ul
				{...passThroughs}
				className={cx(className, '&', {
					'&-is-horizontal': isHorizontal,
					'&-is-vertical': isVertical,
					'&-has-borders': hasBorders,
				})}
			>
				{_.map(itemProps, ({
					hasLine,
					hasPoint,
					pointKind = 1,
					color,
					onClick,
					children,
				}, index) => (
					<li
						key={index}
						className={cx('&-Item')}
						onClick={_.partial(this.handleItemClick, index, itemProps)}
					>
						{hasPoint || hasLine ?
							<svg
								className={cx('&-Item-indicator')}
								width={hasLine || hasSomeLines ? LINE_WIDTH : POINT_SIZE}
								height={POINT_SIZE}
							>
								{hasPoint ?
									<Point
										x={hasLine || hasSomeLines ? LINE_WIDTH / 2 : POINT_SIZE / 2}
										y={POINT_SIZE / 2}
										color={color}
										kind={pointKind}
									/>
								: null}
								{hasLine ?
									<Line
										d={`M0,${POINT_SIZE / 2} L${LINE_WIDTH},${POINT_SIZE / 2}`}
										color={color}
									/>
								: null}
							</svg>
						: null}
						{children}
					</li>
				))}
			</ul>
		);
	},
});

export default Legend;
