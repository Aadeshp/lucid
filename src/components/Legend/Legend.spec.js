import { common } from '../../util/generic-tests';

import Legend from './Legend';

describe('Legend', () => {
	common(Legend);
});
