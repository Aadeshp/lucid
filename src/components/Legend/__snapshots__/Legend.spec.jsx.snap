exports[`Legend [common] example testing should match snapshot for basic 1`] = `
<div>
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="vertical">
    <Legend.Item
      color="color-chart-0"
      hasPoint={true}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasPoint={true}>
      Partner 1
    </Legend.Item>
    <Legend.Item
      color="color-chart-2"
      hasPoint={true}>
      Partner 2
    </Legend.Item>
    <Legend.Item
      color="color-chart-3"
      hasPoint={true}>
      Partner 3
    </Legend.Item>
    <Legend.Item
      color="color-chart-4"
      hasPoint={true}>
      Partner 4
    </Legend.Item>
    <Legend.Item
      color="color-chart-5"
      hasPoint={true}>
      Partner 5
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="vertical">
    <Legend.Item
      color="color-chart-0"
      hasPoint={true}
      pointKind={0}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasPoint={true}
      pointKind={1}>
      Partner 1
    </Legend.Item>
    <Legend.Item
      color="color-chart-2"
      hasPoint={true}
      pointKind={2}>
      Partner 2
    </Legend.Item>
    <Legend.Item
      color="color-chart-3"
      hasPoint={true}
      pointKind={3}>
      Partner 3
    </Legend.Item>
    <Legend.Item
      color="color-chart-4"
      hasPoint={true}
      pointKind={4}>
      Partner 4
    </Legend.Item>
    <Legend.Item
      color="color-chart-5"
      hasPoint={true}
      pointKind={5}>
      Partner 5
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="vertical">
    <Legend.Item
      color="color-chart-0"
      hasLine={true}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasLine={true}>
      Partner 1
    </Legend.Item>
    <Legend.Item
      color="color-chart-2"
      hasLine={true}>
      Partner 2
    </Legend.Item>
    <Legend.Item
      color="color-chart-3"
      hasLine={true}>
      Partner 3
    </Legend.Item>
    <Legend.Item
      color="color-chart-4"
      hasLine={true}>
      Partner 4
    </Legend.Item>
    <Legend.Item
      color="color-chart-5"
      hasLine={true}>
      Partner 5
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="vertical">
    <Legend.Item
      color="color-chart-0"
      hasLine={true}
      hasPoint={true}
      pointKind={0}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasLine={true}
      hasPoint={true}
      pointKind={1}>
      Partner 1
    </Legend.Item>
    <Legend.Item
      color="color-chart-2"
      hasLine={true}
      hasPoint={true}
      pointKind={2}>
      Partner 2
    </Legend.Item>
    <Legend.Item
      color="color-chart-3"
      hasLine={true}
      hasPoint={true}
      pointKind={3}>
      Partner 3
    </Legend.Item>
    <Legend.Item
      color="color-chart-4"
      hasLine={true}
      hasPoint={true}
      pointKind={4}>
      Partner 4
    </Legend.Item>
    <Legend.Item
      color="color-chart-5"
      hasLine={true}
      hasPoint={true}
      pointKind={5}>
      Partner 5
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={true}
    orient="vertical">
    <Legend.Item
      color="color-chart-0"
      hasLine={true}
      hasPoint={true}
      pointKind={0}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasLine={true}
      hasPoint={true}
      pointKind={1}>
      Partner 1
    </Legend.Item>
    <Legend.Item
      color="color-chart-2"
      hasLine={true}
      hasPoint={true}
      pointKind={2}>
      Partner 2
    </Legend.Item>
    <Legend.Item
      color="color-chart-3"
      hasLine={true}
      hasPoint={true}
      pointKind={3}>
      Partner 3
    </Legend.Item>
    <Legend.Item
      color="color-chart-4"
      hasLine={true}
      hasPoint={true}
      pointKind={4}>
      Partner 4
    </Legend.Item>
    <Legend.Item
      color="color-chart-5"
      hasLine={true}
      hasPoint={true}
      pointKind={5}>
      Partner 5
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="horizontal">
    <Legend.Item
      color="color-chart-good"
      hasLine={true}>
      Revenue
    </Legend.Item>
    <Legend.Item
      color="color-chart-bad"
      hasLine={true}>
      Loss
    </Legend.Item>
    <Legend.Item
      color="color-chart-0"
      hasPoint={true}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasPoint={true}>
      Partner 1
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={true}
    orient="horizontal">
    <Legend.Item
      color="color-chart-good"
      hasLine={true}>
      Revenue
    </Legend.Item>
    <Legend.Item
      color="color-chart-bad"
      hasLine={true}>
      Loss
    </Legend.Item>
    <Legend.Item
      color="color-chart-0"
      hasPoint={true}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasPoint={true}>
      Partner 1
    </Legend.Item>
  </Legend>
  <br />
  <Legend
    hasBorders={true}
    isReversed={false}
    orient="vertical">
    <Legend.Item
      color="color-chart-good"
      hasLine={true}>
      Revenue
    </Legend.Item>
    <Legend.Item
      color="color-chart-bad"
      hasLine={true}>
      Loss
    </Legend.Item>
    <Legend.Item
      color="color-chart-0"
      hasPoint={true}>
      Partner 0
    </Legend.Item>
    <Legend.Item
      color="color-chart-1"
      hasPoint={true}>
      Partner 1
    </Legend.Item>
  </Legend>
  <br />
  <br />
  <br />
  <ToolTip
    alignment="end"
    direction="right"
    isExpanded={true}>
    <ToolTip.Target>
      Tooltip example
    </ToolTip.Target>
    <ToolTip.Body>
      <Legend
        hasBorders={false}
        isReversed={false}
        orient="vertical">
        <Legend.Item
          color="color-chart-good"
          hasLine={true}>
          Revenue
        </Legend.Item>
        <Legend.Item
          color="color-chart-bad"
          hasLine={true}>
          Loss
        </Legend.Item>
        <Legend.Item
          color="color-chart-0"
          hasPoint={true}>
          Partner 0
        </Legend.Item>
        <Legend.Item
          color="color-chart-1"
          hasPoint={true}>
          Partner 1
        </Legend.Item>
      </Legend>
    </ToolTip.Body>
  </ToolTip>
</div>
`;
